zerver/tests/test_subs.py

```diff
@@ -719,6 +810,53 @@ def test_bad_delete_parameters(self):
         )
         self.assert_json_error(result, "delete[0] is not a string")
 
+    def test_add_or_delete_not_specified(self):
+        # type: () -> None
+        email = 'hamlet@zulip.com'
+        self.login(email)
+
+        result = self.client_patch(
+            "/api/v1/users/me/subscriptions",
+            {},
+            **self.api_auth(email)
+        )
+        self.assert_json_error(result,
+                               'Nothing to do. Specify at least one of "add" or "delete".')
+
+    def test_invalid_stream_name(self):
+        # type: () -> None
+        email = 'hamlet@zulip.com'
+        self.login(email)
+
+        invalid_stream_name = ""
```

Comment:
Copying @timabbott on this one.

So this method has some really interesting history, which I tracked down in git by iterating a couple times of doing "git blame" then checking out an earlier version of the code.  The way this method evolved was something like this:

* It used to enforce that names started with an alphanumeric: `return re.match("^[.a-zA-Z0-9][.a-z A-Z0-9_-]*$", name`
* For a brief time it always returned True!  (this was probably to minimize refactoring as part of some bigger effort)
* It was then modified to its current state, only enforcing that the string is not empty.
* The method was also moved to another file at one point as part of code reorganization.

@ashishk1994 For a quirky method like this, I would be fine with just adding a little comment to the test that says that the only way to force an error is with the empty string, because `valid_stream_name` is apparently super liberal with what it accepts.  I would consider renaming the test to something more like `test_patch_enforces_valid_stream_name_check`.

We might also want a direct test on this valid_stream_name helper, with something like this:

```
   # valid_stream_name only returns False for empty strings
   self.assertFalse(valid_stream_name('')

  # valid_stream_name allows many types of characters
  self.assertTrue(valid_stream_name('_&*('))
  self.assertTrue(valid_stream_name('   lots of leading/trailing whitespace   '))
  self.assertTrue(valid_stream_name('multiple worlds'))
  self.assertTrue(valid_stream_name('devel'))
  self.assertTrue(valid_stream_name('DeVeL'))